File: go/lens_combinators.go
package efficient

import (
	"fmt"
	"regexp"
	"strings"
)

type RE struct {
	Str   string
	Regex *regexp.Regexp
}

type Lens interface{}

type Del struct {
	RE    RE
	Value string
}

type Store struct {
	RE RE
}

type Value string

type Counter string

type Seq string

type Key struct {
	RE RE
}

type Label string

func concat(l1, l2 Lens, input []string) []string {
	var result []string
	for _, s := range input {
		res1 := applyLens(l1, s)
		for _, r1 := range res1 {
			res2 := applyLens(l2, r1)
			result = append(result, res2...)
		}
	}
	return result
}

func union(l1, l2 Lens, input string) string {
	res1 := applyLens(l1, input)
	if len(res1) > 0 {
		return strings.Join(res1, "")
	}
	return applyLens(l2, input)[0]
}

func repetition(op string, l Lens, input []string) []string {
	var result []string
	for _, s := range input {
		res := applyLens(l, s)
		for _, r := range res {
			result = append(result, r)
		}
		if op == "*" {
			continue
		} else if op == "+" && len(res) == 0 {
			return []string{}
		} else if op == "?" {
			break
		}
	}
	return result
}

func subtree(l Lens, input []string) [][]string {
	var result [][]string
	for _, s := range input {
		res := applyLens(l, s)
		result = append(result, res)
	}
	return result
}

func square(left, body, right Lens, input string) string {
	res1 := applyLens(left, input)
	if len(res1) == 0 {
		return ""
	}
	res2 := applyLens(right, res1[0])
	if len(res2) == 0 {
		return ""
	}
	mid := res2[0][len(res1[0]):]
	midRes := applyLens(body, mid)
	return res1[0] + strings.Join(midRes, "") + res2[0][len(mid):]
}

File: rust/src/lens_combinators.rs
use regex::Regex;

enum RE {
    Str(String),
    Re(Regex),
}

enum Lens {
    Del(RE, String),
    Store(RE),
    Value(String),
    Counter(String),
    Seq(String),
    Key(RE),
    Label(String),
}

fn concat(l1: &Lens, l2: &Lens, input: &[String]) -> Vec<String> {
    let mut result = Vec::new();
    for s in input {
        let res1 = apply_lens(l1, s);
        for r1 in res1 {
            let res2 = apply_lens(l2, &r1);
            result.extend(res2);
        }
    }
    result
}

fn union(l1: &Lens, l2: &Lens, input: &str) -> String {
    let res1 = apply_lens(l1, input);
    if !res1.is_empty() {
        return res1.join("");
    }
    apply_lens(l2, input).join("")
}

fn repetition(op: &str, l: &Lens, input: &[String]) -> Vec<String> {
    let mut result = Vec::new();
    for s in input {
        let res = apply_lens(l, s);
        result.extend(res);
        match op {
            "*" => continue,
            "+" if res.is_empty() => return Vec::new(),
            "?" => break,
            _ => (),
        }
    }
    result
}

fn subtree(l: &Lens, input: &[String]) -> Vec<Vec<String>> {
    let mut result = Vec::new();
    for s in input {
        let res = apply_lens(l, s);
        result.push(res);
    }
    result
}

fn square(left: &Lens, body: &Lens, right: &Lens, input: &str) -> String {
    let res1 = apply_lens(left, input);
    if res1.is_empty() {
        return String::new();
    }
    let res2 = apply_lens(right, &res1[0]);
    if res2.is_empty() {
        return String::new();
    }
    let mid = &res2[0][res1[0].len()..];
    let mid_res = apply_lens(body, mid);
    let mut result = res1[0].clone();
    result.push_str(&mid_res.join(""));
    result.push_str(&res2[0][mid.len()..]);
    result
}